partiql-value/src/lib.rs

```diff
@@ -886,90 +886,97 @@ impl Hash for Bag {
 }
 
 // TODO replace HashMap to support multi key
-#[derive(Default, Eq, Clone)]
-pub struct Tuple(pub HashMap<String, Value>);
+#[derive(Default, Eq, Hash, Debug, Clone)]
+pub struct Tuple {
+    attrs: Vec<String>,
+    vals: Vec<Value>,
+}
+
+impl Tuple {
+    pub fn new() -> Self {
+        Tuple {
+            attrs: vec![],
+            vals: vec![],
+        }
+    }
 
-impl<const N: usize, T> From<[(&str, T); N]> for Tuple
-where
-    T: Into<Value>,
-{
     #[inline]
-    fn from(arr: [(&str, T); N]) -> Self {
-        Tuple(HashMap::from_iter(
-            arr.into_iter().map(|(k, v)| (k.to_string(), v.into())),
-        ))
+    pub fn insert(&mut self, attr: &str, val: Value) {
+        self.attrs.push(attr.to_string());
+        self.vals.push(val);
     }
-}
 
-#[macro_export]
-macro_rules! partiql_tuple {
-    () => (
-         Tuple(HashMap::new())
-    );
-    ($(($x:expr, $y:expr)),+ $(,)?) => (
-        Tuple::from([$(($x, Value::from($y))),+])
-    );
-}
+    #[inline]
+    pub fn get(&self, attr: &str) -> Option<&Value> {
+        match self.attrs.iter().position(|a| *a == attr.to_string()) {
+            Some(i) => Some(&self.vals[i]),
+            _ => None,
+        }
+    }
 
-impl Debug for Tuple {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        match self.0.is_empty() {
-            true => write!(f, "{{}}"),
-            false => {
-                let mut dbg = f.debug_struct("");
-                for (k, v) in &self.0 {
-                    dbg.field(k, v);
-                }
-                dbg.finish()
+    #[inline]
+    pub fn remove(&mut self, attr: &str) -> Option<Value> {
+        match self.attrs.iter().position(|a| *a == attr.to_string()) {
+            Some(i) => {
+                self.attrs.remove(i);
+                Some(self.vals.remove(i))
             }
+            _ => None,
         }
     }
+
+    #[inline]
+    pub fn pairs(&self) -> Vec<(&str, &Value)> {
+        zip(&self.attrs, &self.vals)
+            .map(|(k, v)| (k.as_str(), v))
+            .collect()
+    }
 }
 
-impl PartialOrd for Tuple {
-    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        Some(self.cmp(other))
+impl<const N: usize, T> From<[(&str, T); N]> for Tuple
+where
+    T: Into<Value>,
+{
+    #[inline]
+    fn from(arr: [(&str, T); N]) -> Self {
+        let out = arr
+            .into_iter()
+            .fold(Tuple::new(), |mut acc: Tuple, (attr, val)| {
+                acc.insert(attr, val.into());
+                acc
+            });
+
+        out
     }
 }
 
-impl Ord for Tuple {
```

Comment:
Why is `Ord` removed?

Reply:
Was a miss, fixed in the new commit.

Reply:
Re-visiting this to ensure it conforms with the spec. if so, will re-request a review, thanks.